Make Cargo share more things between servo and ports

nox opened the pull request:
This should enable Cargo to reuse more build artifacts between servo and geckolib.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/9886)

<!-- Reviewable:end -->

nox committed:
Use the unstable feature of string_cache and selectors in geckolib

This should enable Cargo to reuse more build artifacts between servo and geckolib.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/Cargo.toml

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit b3a4348 with merge 54fdfe6...

nox committed:
Use release_max_level_info in all ports

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit a89cf68 with merge b9879e2...

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 71dfeca with merge b6ed5f9...

nox committed:
Enable nightly feature for serde in style

Some external dependencies (at least webrender_traits) present only in servo
and not in geckolib use it already and this prevents Cargo to reuse build
artifacts depending on serde between the two targets.

nox committed:
Enable query_encoding in geckolib

This allows the reuse of teh build artifacts of url and all the things that
depends on it between servo and geckolib.

nox commented:
Last commit probably requires @bholley's seal of approval.

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 949c1a6 with merge 1084c26...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/1939)

nox commented:
Never mind, encoding was already depended upon by cssparser so it's already in geckolib.

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 03d525e with merge 6c1369c...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2220)

nox commented:
Result of running `./mach build-cef --release` after `./mach build --release`:

```
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for embedding v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/ports/cef): failed to read `/home/servo/buildbot/slave/linux-rel/build/target/release/.fingerprint/embedding-73b03625198d6678/lib-embedding`
INFO:cargo::ops::cargo_rustc::fingerprint:   cause: No such file or directory (os error 2)
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for net_traits v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/net_traits): new (msg v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/msg)) != old (msg v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/msg))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for msg v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/msg): new (layers v0.2.2 (https://github.com/servo/rust-layers#dd28dbee)) != old (layers v0.2.2 (https://github.com/servo/rust-layers#dd28dbee))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for layers v0.2.2 (https://github.com/servo/rust-layers#dd28dbee): new (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8)) != old (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for servo-skia v0.20130412.6: new (servo-glutin v0.4.9) != old (servo-glutin v0.4.9)
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for servo-glutin v0.4.9: features have changed: Some(["default"]) != Some(["default", "headless"])
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for servo-glutin v0.4.9: features have changed: Some(["default"]) != Some(["default", "headless"])
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8): new (servo-skia v0.20130412.6) != old (servo-skia v0.20130412.6)
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for script v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/script): new (canvas v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/canvas)) != old (canvas v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/canvas))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for script_traits v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/script_traits): new (canvas_traits v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/canvas_traits)) != old (canvas_traits v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/canvas_traits))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for devtools_traits v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/devtools_traits): new (msg v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/msg)) != old (msg v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/msg))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for canvas_traits v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/canvas_traits): new (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8)) != old (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for gfx_traits v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/gfx_traits): new (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8)) != old (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for canvas v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/canvas): new (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8)) != old (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for compositing v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/compositing): new (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8)) != old (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for gfx v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/gfx): new (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8)) != old (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for layout_traits v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/layout_traits): new (gfx v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/gfx)) != old (gfx v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/gfx))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for glutin_app v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/ports/glutin): features have changed: Some(["window"]) != Some(["headless", "servo-glutin"])
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for devtools v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/devtools): new (devtools_traits v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/devtools_traits)) != old (devtools_traits v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/devtools_traits))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for servo v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/servo): features have changed: Some(["default", "glutin_app", "webdriver", "webdriver_server", "window"]) != Some(["glutin_app", "headless"])
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for net v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/net): new (devtools_traits v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/devtools_traits)) != old (devtools_traits v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/devtools_traits))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for layout v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/layout): new (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8)) != old (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for webdriver_server v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/webdriver_server): failed to read `/home/servo/buildbot/slave/linux-rel/build/target/release/.fingerprint/webdriver_server-25f279ca64df6ed4/lib-webdriver_server`
INFO:cargo::ops::cargo_rustc::fingerprint:   cause: No such file or directory (os error 2)
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for webdriver v0.4.0 (https://github.com/jgraham/webdriver-rust.git#9dffcbe4): failed to read `/home/servo/buildbot/slave/linux-rel/build/target/release/.fingerprint/webdriver-b7a550d30588e536/lib-webdriver`
INFO:cargo::ops::cargo_rustc::fingerprint:   cause: No such file or directory (os error 2)
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for servo v0.0.1 (file:///home/servo/buildbot/slave/linux-rel/build/components/servo): features have changed: Some(["default", "glutin_app", "webdriver", "webdriver_server", "window"]) != Some(["glutin_app", "headless"])
```

nox commented:
I think this all comes from the headless feature, and we could alleviate some of it if we get #8641 in.

SimonSapin commented:
r-

This doesn’t matter at all. When multiple crates depend on, for example, `string_cache` (with version requirements that are semver-compatible with each other) Cargo will only build the crate once and use the union of all optional features that at least one user requires. There is no need to repeat these lists of features everywhere. In fact, repeating them will give us more work when we’ll want to update them in the future.

SimonSapin commented:
In fact, I’d be in favor of going full "Don’t Repeat Yourself" and only listing optional features and version requirements in `Cargo.toml` for top-level crates, with `[dependencies] something = "*"` in "intermediate" crates of this repository.

nox commented:
I toned down the first commit, but I can't do that for the serde one given serde does not appear explicitly in ports/geckolib/Cargo.toml.

nox commented:
@SimonSapin Some need to be repeated otherwise they don't get build the same in all four toplevel crates.

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 9a358ab with merge 59b7404...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2221)

SimonSapin commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit e6d4f5c has been approved by `SimonSapin`

<!-- @bors-servo r=SimonSapin e6d4f5cad8c4c2bca80518d46b7c1feac2117fbb -->

bors-servo commented:
:hourglass: Testing commit e6d4f5c with merge fc9dcc4...

bors-servo commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/servo/servo/build/1.0.86)

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit d59cad2 with merge d344f33...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1934), [gonk](http://build.servo.org/builders/gonk/builds/1980), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1956), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2237), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/655), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2008), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2062), [status-appveyor](https://ci.appveyor.com/project/servo/servo/build/1.0.108)

Wafflespeanut commented:
@bors-servo r=SimonSapin

bors-servo commented:
:pushpin: Commit d59cad2 has been approved by `SimonSapin`

<!-- @bors-servo r=SimonSapin d59cad2dd9aace8a7a7f67facd5076e4d83117e7 -->

bors-servo commented:
:hourglass: Testing commit d59cad2 with merge 5fbed88...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1946), [gonk](http://build.servo.org/builders/gonk/builds/1992), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1968), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2249), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/667), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2020), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2074), [status-appveyor](https://ci.appveyor.com/project/servo/servo/build/1.0.143)
